go/src/lib.rs

```diff
@@ -1,9 +1,172 @@
 /**
  * Copyright GLIDE-for-Redis Project Contributors - SPDX Identifier: Apache-2.0
  */
-use std::ffi::c_void;
+use glide_core::client::Client as GlideClient;
+use glide_core::connection_request;
+use protobuf::Message;
+use std::{
+    ffi::{c_void, CString},
+    os::raw::c_char,
+};
+use tokio::runtime::Builder;
+use tokio::runtime::Runtime;
 
+/// Success callback that is called when a Redis command succeeds.
+pub type SuccessCallback =
+    unsafe extern "C" fn(channel_address: usize, message: *const c_char) -> ();
+
+/// Failure callback that is called when a Redis command fails.
+///
+/// `error` should be manually freed by calling `free_error` after this callback is invoked, otherwise a memory leak will occur.
+pub type FailureCallback =
+    unsafe extern "C" fn(channel_address: usize, error: *const RedisErrorFFI) -> ();
+
+/// The connection response.
+///
+/// It contains either a connection or an error. It is represented as a struct instead of an enum for ease of use in the wrapper language.
+#[repr(C)]
+pub struct ConnectionResponse {
+    conn_ptr: *const c_void,
+    error: *const RedisErrorFFI,
+}
+
+/// A Redis error.
+#[repr(C)]
+pub struct RedisErrorFFI {
+    message: *const c_char,
+    error_type: ErrorType,
+}
+
+/// FFI compatible version of the ErrorType enum defined in protobuf.
+// TODO: Update when command errors are implemented
+#[repr(C)]
+pub enum ErrorType {
+    ClosingError = 0,
+    RequestError = 1,
+    TimeoutError = 2,
+    ExecAbortError = 3,
+    ConnectionError = 4,
+}
+
+/// The glide client.
+#[allow(dead_code)]
+pub struct Client {
+    client: GlideClient,
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+    runtime: Runtime,
+}
+
+fn create_client_internal(
+    connection_request_bytes: &[u8],
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+) -> Result<Client, RedisErrorFFI> {
+    let request = connection_request::ConnectionRequest::parse_from_bytes(connection_request_bytes)
+        .map_err(|err| {
+            let msg = CString::new(err.to_string()).unwrap();
+            RedisErrorFFI {
+                message: msg.into_raw(),
+                error_type: ErrorType::ConnectionError,
```

Comment:
The Java code has a test that connects to an invalid address and then asserts that a ClosingException occurred, should this be ClosingError instead of ConnectionError?
https://github.com/Bit-Quill/glide-for-redis/blob/main/java/integTest/src/test/java/glide/ErrorHandlingTests.java#L34

Reply:
@shachlanAmazon said no in https://github.com/aws/glide-for-redis/pull/1077#discussion_r1514297208, bit I hope he will change his mind later.
We can keep it as `ConnectionError` and change later